lib/ddtrace/utils.rb

```diff
@@ -29,8 +33,24 @@ def self.truncate(value, size, omission = '...')
       string.slice(0, size - omission.size) + omission
     end
 
-    private_class_method :reset!, :was_forked?
+    def utf8_encode(str, options = {})
+      str = str.to_s
+
+      if str.encoding?(Encoding::UTF_8)
+        str
+      elsif options[:binary]
+        # This option is useful for "gracefully" displaying binary data that
+        # often contains text such as marshalled objects
+        str.encode('UTF-8', 'binary', invalid: :replace, undef: :replace, replace: '')
```

Comment:
Non-blocking comment.

I think as its written, this is readable and acceptable.

However, I would suggest that perhaps, given the different combinations of encodings and options, the responsibilites of `#utf8_encode` is actually broader than it looks, and that binary encoding itself might warrant its own function to simplify `#utf8_encode`'s responsibilities.

Breaking binary encoding out into its own function would provide the benefit of directly testing binary encoding itself via unit tests. I do realize the proposed function would be one line, and likely not contribute anything to the readability of this already satisfactory code, so I defer to your judgment whether you think this separation of responsibility is warranted by its benefits or not.

Reply:
One of the reasons we're experiencing encoding issues is that we often resort to different approaches on how to solve this. I don't think we should have more than one method in the public API doing that.

Considering that, I don't understand what's the gain of having another method for testing purposes as opposed to assertions with `binary: true` options. Both can provide the same amount of test granularity.